packages/agents/src/index.ts

```
@@ -1458,32 +1461,107 @@ export class Agent<
   }
 
+  /**
+   * Restore MCP server state from database for OAuth callbacks after DO hibernation.
+   * Checks if request is an OAuth callback and restores the callback URL and connection if needed.
+   */
+  private async _restoreMcpStateForOAuthCallback(
+    request: Request
+  ): Promise<void> {
+    // Check if this is an OAuth callback by URL pattern
+    // Pattern: /agents/{agent-name}/{agent-id}/callback/{serverId}?code=...
+    const url = new URL(request.url);
+    const isCallbackPath =
+      url.pathname.includes("/callback/") && url.searchParams.has("code");
+
+    if (!isCallbackPath) {
+      return;
+    }
+
+    // Extract serverId from callback URL
+    const pathParts = url.pathname.split("/");
+    const callbackIndex = pathParts.indexOf("callback");
+    const serverId = callbackIndex !== -1 ? pathParts[callbackIndex + 1] : null;
+
+    if (!serverId) {
+      return;
+    }
+
+    // Restore callback URLs and connection from database if not in memory
+    const server = this.sql<MCPServerRow>`
+      SELECT id, name, server_url, client_id, auth_url, callback_url, server_options
+      FROM cf_agents_mcp_servers
+      WHERE id = ${serverId}
+    `.find((s) => s.id === serverId);
+
+    if (!server) {
+      return;
+    }
+
+    // Register callback URL (restores it after hibernation)
+    if (server.callback_url) {
+      this.mcp.registerCallbackUrl(`${server.callback_url}/${server.id}`);
+    }
+
+    // Restore connection if not in memory
+    if (!this.mcp.mcpConnections[serverId]) {
+      try {
+        await this._connectToMcpServerInternal(
+          server.name,
+          server.server_url,
+          server.callback_url,
+          server.server_options ? JSON.parse(server.server_options) : undefined,
+          {
+            id: server.id,
+            oauthClientId: server.client_id ?? undefined
+          }
+        );
+      } catch (error) {
+        console.error(`Failed to restore connection for ${serverId}:`, error);
```

hmm should it be a silent error? what can we do better here?

---

do better